internal/app/singularity/verify_ocsp.go

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2022, Sylabs Inc. All rights reserved.
+// Copyright (c) 2020-202, ICS-FORTH.  All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the LICENSE.md file
+// distributed with the sources of this project regarding your rights to use or distribute this
+// software.
+
+package singularity
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/sylabs/singularity/pkg/sylog"
+	"golang.org/x/crypto/ocsp"
+)
+
+/*
+Online Certificate Status Protocol - OCSP
+
+OCSP responder is used to provide real-time verification of the revocation status of an X.509 certificate.
+RFC: https://www.rfc-editor.org/rfc/rfc6960
+*/
+
+const (
+	PKIXOCSPNoCheck = "1.3.6.1.5.5.7.48.1.5"
+)
+
+var errOCSP = errors.New("OCSP verification has failed")
+
+func OCSPVerify(chain ...*x509.Certificate) error {
+	// use the pool as an index for certificate issuers.
+	// fixme: we can drop this lookup if we assume that certificate N is always signed by certificate N+1.
+	pool := map[string]*x509.Certificate{}
+
+	for _, cert := range chain {
+		pool[string(cert.SubjectKeyId)] = cert
+	}
+
+	// recursively validate the certificate chain
+	for _, cert := range chain {
+		if err := validateCertificate(cert, pool); err != nil {
+			sylog.Warningf("OCSP verification has failed. Err: %s", err)
+			return errOCSP
+		}
+	}
+
+	return nil
+}
+
+func validateCertificate(cert *x509.Certificate, pool map[string]*x509.Certificate) error {
+	if len(cert.AuthorityKeyId) == 0 || string(cert.SubjectKeyId) == string(cert.AuthorityKeyId) {
+		sylog.Infof("skip self-signed certificate (%s)", cert.Subject.String())
+
+		return nil
+	}
+
+	/*---------------------------------------------------
+	 * Retrieve the CA who issued the certificate in question.
+	 *---------------------------------------------------*/
+
+	// firstly, look for the issuer in the pool of certificates.
+	issuer, exists := pool[string(cert.AuthorityKeyId)]
+	if !exists {
+		// if not found anywhere locally, try to download it
+		missingCerts, err := downloadCertsFromURLs(cert.IssuingCertificateURL...)
+		if err != nil {
+			return errors.Wrapf(err, "download cert error")
+		}
+
+		// if that does not work either, just abort
+		issuer, exists = missingCerts[string(cert.AuthorityKeyId)]
+		if !exists {
+			return errors.Errorf("cannot find issuer '%s'", cert.Issuer)
+		}
+	}
```

Comment:
By definition, certificate N in the chain was issued by certificate N+1 all the way to the root, so I don't think we should ever need to download a certificate during OCSP verification?

Reply:
This part is used to validate the OCSP's certificate -- not for the code signing certificate. 

At least, this is my interpretation of `4.2.2.2  Authorized Responders` on https://www.ietf.org/rfc/rfc2560.txt

Reply:
> This part is used to validate the OCSP's certificate -- not for the code signing certificate.

I think it's used for both? It's called from `OCSPVerify` [here](https://github.com/fnikolai/singularity-x509/blob/49bb33b0f81c84517d679e18e8324581fa93dd7d/internal/app/singularity/verify_ocsp.go#L48-L54), which is validating the code signing certificate?

> At least, this is my interpretation of `4.2.2.2 Authorized Responders` on https://www.ietf.org/rfc/rfc2560.txt

My understanding from that section is there are three cases to consider:

1. _Matches a local configuration of OCSP signing authority for the certificate in question:_ In this case, the responder certificate would be contained in the local configuration, and 
2. _Is the certificate of the CA that issued the certificate in question:_ In this case, the responder certificate is the same as the CA of the certificate in question, and included in the chain.
4. _Includes a value of `id-ad-ocspSigning` in an ExtendedKeyUsage extension and is issued by the CA that issued the certificate in question:_ In this case, the responder certificate must be signed by the same CA as the certificate in question, and thus can be validated using the same intermediate and root pools.

In all three cases, it's unnecessary to download an issuer (CA) certificate? It's either in local configuration (we're not implementing this currently), or present as a CA in the chain. Am I missing a case where the issuer of a OCSP certificate needs to be downloaded?

Reply:
I was thinking about the case of an OCSP responder being signed by intermediate `ocspSigning` certificates that have the same Root CA as the `codeSigning` certificates. In that case, we had to download the chain of intermediate `ocspSigning` certificates.

In other words, my assumption was the OCSP Responder can be indirectly signed by the RootCA. But this case is excluded by `2.2  Response`:
```
The key used to sign the response MUST belong to one of the following:
   -- the CA who issued the certificate in question
   -- a Trusted Responder whose public key is trusted by the requester
   -- a CA Designated Responder (Authorized Responder) who holds a specially marked certificate issued directly by the CA, indicating that the responder may issue OCSP responses for that CA.
```

So, I 'm removing the corresponding code.